### PR TITLE
feat: DNS.MaxCacheTTL for DNS-over-HTTPS resolvers

### PR DESCRIPTION
**Pull request by thibmeu**
Expose `WithMaxCacheTTL` option of go-doh-resolver within IPFS node config.
This allows to cap the Time-To-Live suggested by the DNS response. This includes resolution of DNSLink captured by the custom resolver URLs.

This change relies on [go-ipfs.config](https://github.com/ipfs/go-ipfs-config/pull/161) change. The change of go.mod and go.sum would have to be updated accordingly at this time.

Before merging, both go-doh-resolver and go-ipfs-config have to be updated.

Tagging @vyzo and @marten-seemann , as they reviewed the [go-doh-resolver change](https://github.com/libp2p/go-doh-resolver/pull/12).

### CODE REVIEW

File: go.mod
Comment (vyzo) [inline go.mod:68]: this needs tagged version.

File: core/node/dns.go
Comment (vyzo) [inline core/node/dns.go:35]: this will overflow...  do we actually need to set it?

File: go.mod
Comment (thibmeu) [reply go.mod:68]: it does. The tagged version has not been released yet, therefore the mention to the commit directly, and a reference to this update in the PR message.
Release is tracked in https://github.com/libp2p/go-doh-resolver/issues/15

File: core/node/dns.go
Comment (thibmeu) [reply core/node/dns.go:35]: we need to set `WithDefault` to something. I considered using `0`, since there is an if before and we know the option is set.

`math.MaxUint32` is the default in [go-doh-resolver](https://github.com/libp2p/go-doh-resolver/blob/master/resolver.go#L63). We might want to update if there is an overflow.

Comment (vyzo) [reply core/node/dns.go:35]: actually this is fine, duration is 64 bit and int32 max is not that big in that case.

File: go.mod
Comment (vyzo) [reply go.mod:68]: yeah, just putting an anchor to replace with release tag before merging.

Comment (thibmeu) [reply go.mod:68]: The change is now included in `libp2p/go-doh-resolver@0.4.0`.
go-ipfs-config update is likely to be in `ipfs/go-ipfs-config@0.19.0`.

@aschmahmann do you know if there are other changes to be made to go-ipfs-config for go-ipfs 0.13 release? If not, I can update this commit to have the correct version and hash in `go.mod` and `go.sum`.

**Review by lidel — CHANGES_REQUESTED** [review]
@thibmeu mind elaborating why you want to cap the Time-To-Live suggested by the DNS response? What about being able to set the minimum, or override TTL for IPNS records as well?

FYSA I am looking into adding a proper `cache-control` header to all HTTP responses under  `/ipns/`  namespace (https://github.com/ipfs/go-ipfs/issues/1818#issuecomment-1015849462) – Gateway is not utilizing TTL from DNS at all right now. This means your PR will  not have much impact on gateway responses, but maybe you have a different goal?

My understanding is we want to fix cache control in Gateway responses for `/ipns/`, and need to wait for #1818. After we wire up TTLs, and have both `cache-control` and `etag` in Gateway responses for `/ipns/` content paths, then we can pick this PR up and tweak  `cache-control`  / based on TTL minimum and maximum for both IPNS and DNSLink records. 

I believe we need to have config option for both min and max, because there are people setting records with very small TTL such as 60 seconds, and Gateway operators should be able to override the minimum to be higher, to leverage the cache more.

If we want to control this at the record cache level, then we need four OptionalDurations:
- min/max for DNSLink (and multiaddrs which resolve DNS)
  - `DNS.MaxCacheTTL`
  - `DNS.MinCacheTTL`
- min/max for IPNS records
  - `Ipns.MaxCacheTTL`
  - `Ipns.MinCacheTTL`

Instead, we could introduce min/max specific to Gateway.
It would be applied at the very end, when `cache-control` header is calculated for both DNSLink and IPNS records:
 - `Gateway.MaxCacheTTL`
 - `Gateway.MinCacheTTL`

Of course, `/ipfs/` would remain immutable, would be applied only to mutable responses under `/ipns/` (including DNSLink websites).

I am leaning towards  adding `Gateway.*` ones for now, but lmk your thoughts and use case, perhaps there is a better/simpler way for controlling this?

File: go.mod
Comment (lidel) [inline go.mod:34]: :information_source:  (note to self) we need to switch to release version before merging

**Review by lidel — APPROVED** [review]
Thank you @thibmeu,  in this context it makes sense  to move forward with this PR: it is not directly related to https://github.com/ipfs/go-ipfs/issues/1818 and brings value on its own. 

Pushed some cosmetic changes:
- rebased to fix some CI issues
- switched to release version of go-doh-resolver@v0.4.0 and go-ipfs-config from upstream master (merged https://github.com/ipfs/go-ipfs-config/pull/161)
- tweaked docs:  given that we don't control cache of the default (OS-level) resolver, made it clear that this setting requires DoH
- fixed broken DoH examples while at it


LGTM and ready to merge, but parking until we tag a new release of `go-ipfs-config`.

@aschmahmann are you aware of any other config changes in 0.13 that we should wait for? 
If not, I'd like to tag and merge this to clear up the queue.

**Comment by aschmahmann** [comment]
> @aschmahmann are you aware of any other config changes in 0.13

Yes, there will likely be some related to the next go-libp2p release

>  that we should wait for?

🤷. It doesn't cost us anything to cut more go-ipfs-config releases so we can go ahead and merge + release. I don't know how many people are running builds from master though/if it really makes a difference though. Certainly if you think that we'll have to worry about rebasing the PRs and dealing with conflicts here then definitely just tag + merge.

**Review by lidel — APPROVED** [review]
Switched to go-ipfs-config v0.19.0, this is ready for merge.

**Comment by BigLep** [comment]
Next step: review the help text.

File: docs/config.md
Comment (guseggert) [inline docs/config.md:1800]:
```suggestion
Maximum duration that entries are valid in the DoH cache.
```

File: docs/config.md
Comment (guseggert) [inline docs/config.md:1802]:
```suggestion
This allows you to cap the Time-To-Live suggested by the DNS response ([RFC2181](https://datatracker.ietf.org/doc/html/rfc2181#section-8)).
```

File: docs/config.md
Comment (guseggert) [inline docs/config.md:1805]:
```suggestion
Note: this does NOT work with Go's default DNS resolver. To make this a global setting, add a `.` entry to `DNS.Resolvers` first.
```

File: docs/config.md
Comment (guseggert) [inline docs/config.md:1813]: For completeness, can we add this to the `Types` section?

Comment (lidel) [reply docs/config.md:1813]: Good idea, but opened https://github.com/ipfs/go-ipfs/pull/8729 to not block this PR.

**Pull request merged**
